feat(flags): Pro active caching of local eval flag info

benjackwhite opened the pull request:
## Problem

We want to ditch the standard API model for loading flags which is unnecessarily hard on our DB. We also want to guarantee as many 9s as possible of uptime do decoupling it entirely from our other processes.

This is step 1 of that - pre-warming a hot cache which will be a mixture of S3 and Redis. S3 will be the new "source of truth" which we update via background tasks in Django. This won't be public and so can contain a lot of info (this is _almost_ what we do with the RemoteConfig concept (follow up will add s3 writes to that too).

The API itself has now a GET option to use the cache which will first hit redis, then S3 and only after that PG. This is just for testing - the final service (rust most likely) will just do Redis->S3.

This doesn't implement the pro active cache of token validation which will come later

## Changes

- [x] Adds a new `local_evaluation.py` which contains a copy of all the methods in the API code (once we are happy this works we can remove the duplicated code
- [x] Adds the logic for automatically updating the cache of info. It also fully replace all caches once per day (we can discuss whether we really need this

## How did you test this code?

- [x] Some tests but im not the flags expert so its minimal atm

## Follow up
- [ ] do some testing to ensure the files content all looks correct and that the API responds with the new param as expected
- [ ] Make a dashboard like we have for remote config
- [ ] Follow up PR to remove the old code to only use the new path
- [ ] Follow up PR doing the same pre-caching for token validation logic

benjackwhite committed:
Fix caching of local eval

benjackwhite committed:
Added local eval model builder

benjackwhite committed:
Added cache checks everywhere

benjackwhite committed:
fixes

benjackwhite committed:
Fixes

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

benjackwhite committed:
Fixes

benjackwhite committed:
Merge branch 'feat/hyper-cache-remote-config-flags' of github.com:PostHog/posthog into feat/hyper-cache-remote-config-flags

benjackwhite committed:
fixes

benjackwhite committed:
Fixes

github-actions[bot] committed:
Update query snapshots

benjackwhite committed:
Added hypercache

benjackwhite committed:
Fixes

benjackwhite committed:
Added hypercache to remoteconfig

benjackwhite committed:
Added tests

benjackwhite committed:
fixes

benjackwhite committed:
Merge branch 'feat/hyper-cache-flags-2' into feat/hyper-cache-remote-config-flags

benjackwhite committed:
Merge branch 'master' into feat/hyper-cache-remote-config-flags

benjackwhite committed:
fix

benjackwhite committed:
Fixes

benjackwhite committed:
Fixes

benjackwhite committed:
fixes

benjackwhite committed:
Fix flags

benjackwhite committed:
fix local dev

benjackwhite committed:
fix

benjackwhite committed:
fix

benjackwhite committed:
Fixes

benjackwhite committed:
Fixes

benjackwhite committed:
fix

benjackwhite committed:
type fixes

benjackwhite committed:
fix

benjackwhite committed:
fixes

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

benjackwhite committed:
Fixes

benjackwhite committed:
Merge branch 'feat/hyper-cache-remote-config-flags' of github.com:PostHog/posthog into feat/hyper-cache-remote-config-flags

benjackwhite committed:
fixes

benjackwhite committed:
fixes

benjackwhite committed:
fixes

github-actions[bot] committed:
Update query snapshots

benjackwhite committed:
Fixes

benjackwhite committed:
Merge branch 'feat/hyper-cache-remote-config-flags' of github.com:PostHog/posthog into feat/hyper-cache-remote-config-flags

benjackwhite committed:
Fixes

benjackwhite committed:
fix

github-actions[bot] committed:
Update query snapshots

benjackwhite committed:
Fixes

benjackwhite committed:
Merge branch 'feat/hyper-cache-remote-config-flags' of github.com:PostHog/posthog into feat/hyper-cache-remote-config-flags

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

benjackwhite committed:
Merge branch 'master' into feat/hyper-cache-remote-config-flags

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

benjackwhite committed:
Merge branch 'master' into feat/hyper-cache-remote-config-flags

benjackwhite committed:
Fixes

benjackwhite committed:
fix

benjackwhite committed:
fix

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

haacked commented on posthog/models/feature_flag/local_evaluation.py:
```diff
@@ -0,0 +1,174 @@
+            else:
+                feature_flag.filters = filters
+
+            cohort_ids = feature_flag.get_cohort_ids(
```
When I tested this, the `cohorts` section was empty when I added `send_cohorts&use_cache` to the query string.

The cohort_ids are captured after the flag transformation, but the transformation removes cohort references. This causes the cohorts dict to be empty even when flags use cohorts.

Fix: Move `cohort_ids = feature_flag.get_cohort_ids(...)` to before the transformation (around line 81).

haacked replied:
Ok, I pushed my fix.

haacked reviewed (changes requested):
Excited about these changes! A few things though:

1. Need to have a signal for `GroupTypeMapping` because changes to those will affect the flag definitions.
2. Probably should add some integration tests of `local_evaluation` where we set `use_cache` in the query string.
3. Found a bug in porting the local eval code where `send_cohorts&use_cache` was not sending the cohorts.

haacked committed:
Add integration tests for `use_cache` parameter in `local_evaluation`

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

haacked committed:
Fix `send_cohort` behavior when `use_cache` is set

The problem is `send_cohorts` is a little confusing. We always send cohorts as part of the response, but if `send_cohorts=true` we send cohort definitions in a top-level cohorts field.

If `send_cohorts=false`, we transform the filters for the flags that have cohort filters to include the cohort conditions. But this only supports simple person property filters.

haacked committed:
Fix test

haacked committed:
feat: add GroupTypeMapping signal handler for HyperCache invalidation

Add missing signal handlers to invalidate HyperCache when GroupTypeMapping
changes, ensuring cache consistency for flag definitions that include
group type mappings.

Changes:
- Add post_save and post_delete signal handlers for GroupTypeMapping
- Handle project_id to team_id mapping since GroupTypeMapping uses project_id
- Add comprehensive test coverage for both update and delete scenarios
- Test specifically targets HyperCache system (use_cache parameter) vs old cache

This addresses the cache invalidation gap identified in PR review where
GroupTypeMapping changes weren't invalidating the new HyperCache system,
only the legacy FlagDefinitionsCache.

haacked committed:
Add `post_delete` receivers to be on the safe side

haacked commented on posthog/models/feature_flag/local_evaluation.py:
```diff
@@ -0,0 +1,174 @@
+# NOTE: All models that affect the cache should have a signal to update the cache
+
+
+@receiver(post_save, sender=FeatureFlag)
```
I added the post_delete receivers too. Let me know if that was unnecessary or even incorrect.

benjackwhite replied:
yeah makes sense although this would also be covered by the daily sync (the code is there, just not the scheduled task which I'll add in follow up)

haacked committed:
feat: add comprehensive post_delete signal handlers and tests

Fix signal handler signatures to handle both post_save and post_delete
signals properly, and add comprehensive test coverage for cache
invalidation on model deletion.

Changes:
- Fix signal handler signatures to use **kwargs instead of explicit created parameter
- Rename functions for clarity: feature_flag_saved -> feature_flag_changed, cohort_saved -> cohort_changed
- Add test for FeatureFlag deletion cache invalidation
- Add test for Cohort deletion cache invalidation

This ensures the HyperCache system properly invalidates when any of the
core models (FeatureFlag, Cohort, GroupTypeMapping) are deleted, maintaining
cache consistency.

haacked commented:
> It also fully replace all caches once per day (we can discuss whether we really need this

I didn't see this in the PR, but I don't think this PR needs it right now.

haacked committed:
Merge remote-tracking branch 'origin/master' into feat/hyper-cache-remote-config-flags

haacked committed:
Update test snapshots after merge

haacked reviewed (approved):
Ok, I fixed the issues. I'd like another pair of eyes on it though.

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

pauldambra reviewed (approved):
stamping to remove codeowners review request

`/* cohort_calculation: */` changed in the replay `ambr` files

latest codeowners in master only cares about `.py` files

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

haacked commented:
I tested the output from before and after this change (with `use_cache`) and confirmed that the output is the same and that the files are stored in the S3 simulator correctly.

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

haacked committed:
fix: make API tokens deterministic in organization feature flag tests

This fixes the infinite loop of 'Update query snapshots' commits caused by
non-deterministic API tokens in test_organization_feature_flag.py tests.

The @snapshot_postgres_queries decorator was capturing queries that included
randomly generated api_token values (like 'phc_w0CoAHAejrzaIMZm68AzQ25euuAz9WdxAidjmTchNMS').
Each test run generated different tokens, causing snapshots to differ and
triggering endless snapshot update commits in CI.

Changes:
- Set deterministic API tokens in both test class setUp methods
- Updated snapshots to use the new stable token values
- Ensures all Team objects created during tests have predictable tokens

This follows the same pattern used in test_remote_config.py which sets
'self.team.api_token = "phc_12345"' for snapshot stability.

haacked commented on posthog/models/feature_flag/local_evaluation.py:
```diff
@@ -0,0 +1,199 @@
+
+@receiver(post_save, sender=GroupTypeMapping)
+@receiver(post_delete, sender=GroupTypeMapping)
+def group_type_mapping_changed(sender, instance: "GroupTypeMapping", created=None, **kwargs):
```
Should we delete the block then?
